// wdio-wiremock-service/lib/plugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plugin = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
class Plugin {
    constructor(options = {}, capabilities, config) {
        const args = options.args || [];
        const version = options.version || '2.27.2';
        const rootDir = options.rootDir || 'wiremock';
        if (!(0, fs_1.existsSync)(rootDir)) {
            (0, fs_1.mkdirSync)(rootDir, { recursive: true });
        }
        if (args.includes('-port')) {
            throw new Error('Cannot set port using args. Use options.port instead.');
        }
        if (args.includes('-root-dir')) {
            throw new Error('Cannot set root-dir using args. Use options.rootDir instead.');
        }
        this.silent = !!options.silent;
        this.watchMode = !!(config === null || config === void 0 ? void 0 : config.watch);
        this.binPath = (0, path_1.resolve)(__dirname, `wiremock-standalone-${version}.jar`);
        this.port = options.port || 8080;
        this.skipWiremockInstall = !!options.skipWiremockInstall;
        this.args = ['-jar', this.binPath, '-port', this.port.toString(), '-root-dir', rootDir].concat(args);
        this.url =
            `${options.mavenBaseUrl || 'https://repo1.maven.org/maven2'}` +
                `/com/github/tomakehurst/wiremock-standalone/${version}/wiremock-standalone-${version}.jar`;
    }
}
exports.Plugin = Plugin;
